### DISCUSSION TITLE
Supporting macro expand for rust.

### DISCUSSION THREAD

**Discussion by YHM404** [helix-editor/helix#6166] [Q&A]
When I use vscode, I can expand a rust macro like this [example](https://rust-analyzer.github.io/manual.html#expand-macro-recursively).

How can I do the same things in Helix?

**Comment by the-mikedavis** [comment]
This is implemented in rust-analyzer as an extension to LSP https://github.com/rust-lang/rust-analyzer/blob/a8d3c46082696b5f7dcf4b960200f2cdc2e48bd8/crates/rust-analyzer/src/lsp_ext.rs

LSP extensions aren't supported currently and should most likely wait for a plugin system. So a plugin system exists and someone creates a plugin to implement rust-analyzer's LSP extensions, macro expansion will not be usable with Helix

**Reply by pascalkuthe** [reply]
I am sort of confused why tough. It *should* be possible to create this with an LSP command that either:

* writes the result to a temporary file that is opened with `window/showDocument`
* replaces the macro with its expansion (potentially separate command)

I might look into submitting a PR for that upstream since RA macro expansion is quite useful

**Comment by Velnbur** [comment]
This feature is one of the parts, that I'm really missing in helix. Is there a workaround for it for now?

**Reply by senekor** [reply]
[cargo-expand](https://github.com/dtolnay/cargo-expand) is a good workaround. You have to leave the editor and manually specify what you want to expand, so it's not quite as convenient. But you can specify very precisely what to expand based on module / item paths, so it does get the job done.

**Comment by irh** [comment]
I got the following to work with Steel support enabled, it calls `rust-analyzer/expandMacro` and  places the expanded macro in a `vsplit` scratch buffer:

```scheme
(require "helix/editor.scm")
(require "helix/commands.scm")
(require "helix/configuration.scm")
(require "helix/misc.scm")
(require "helix/static.scm")
(require-builtin helix/core/text as text.)

(provide shell rust-analyzer-expand-macro)

(define (current-path)
  (let* ([focus (editor-focus)]
         [focus-doc-id (editor->doc-id focus)])
    (editor-document->path focus-doc-id)))

(define (focused-document->rope)
  (let* ([focus (editor-focus)]
         [focus-doc-id (editor->doc-id focus)])
    (editor->text focus-doc-id)))

(define (current-document-cursor-line)
  (define rope (focused-document->rope))
  (define pos (cursor-position))
  (text.rope-char->line rope (cursor-position)))

(define (current-document-cursor-column)
  (define rope (focused-document->rope))
  (define pos (cursor-position))
  (define line (text.rope-char->line rope (cursor-position)))
  (define line-start (text.rope-line->char rope line))
  (- pos line-start))

;;@doc
;; Expands the macro under the cursor
(define (rust-analyzer-expand-macro)
  (send-lsp-command "rust-analyzer"
    "rust-analyzer/expandMacro"
    (hash
      "textDocument" (hash "uri" (string-append "file://" (current-path)))
      "position" (hash
        "line" (current-document-cursor-line)
        "character" (current-document-cursor-column)))
    (lambda (result)
      (cond
        ((void? result) (set-error! "No macro found under cursor"))
        ((hash? result) (begin
                          (define name (hash-get result 'name))
                          (define expansion (hash-get result 'expansion))
                          (vsplit-new)
                          (set-scratch-buffer-name! (string-append "[" name "]"))
                          (set-language "rust")
                          (insert_string expansion)
                          (goto_file_start)))))))
```
